### ISSUE TITLE
Overwriting official commits, metadata

### ISSUE THREAD

**Issue by rousskov** [measurement-factory/bzr2git4notes#5]
Please check whether the following two importing rules solve all the branch nick problems we have discussed without reopening #4:

1. A being-imported branch never overwrites any existing commits and existing annotations.
2. When importing a branch, all new annotations use the git branch name of that branch.

If we want to import _merged_ feature branches into the same repository, we could add the following rule:

3. Iff the revision number of a being imported bzr commit differs from the revision number(s) in the existing annotation for that commit, then the importer _adds_ another `Bzr-Reference` line to that existing annotation:

```
Bzr-Reference: v5 r14884.1.45
Bzr-Reference: gnutls-pt2 r14930
```

However, that third rule is probably difficult for you to implement and, more importantly, it may result in polluting official annotations with annotations from developer repositories if a developer accidentally pushes her annotations into the official repository. While pushing the wrong branch into the official repository may happen as well, it feels much easier to fix that than fixing a wrongly-combined annotation.

Official code annotations are official metadata and should be controlled by the Project. Merging Project and non-Project metadata is clearly a bad idea. Thus, I have rejected that third rule. Non-Project code can still use annotations, of course, but they cannot combine them with official ones in one git object.

As a consequence of the above two rules, we will need to import branches in the "right order", but I think we can manage that without much effort for official branches. Earlier, I was worried about bagX and similar unofficial branches, but the "never overwrite" rule will protect the official commits in the repository and any annotation conflicts among bagX branches is a much smaller problem that we may be able to ignore or deal with on a case-by-case/manual basis.

**Comment by eduard-bagdasaryan** [comment]
To my surprise, [7608ade](https://github.com/measurement-factory/bzr2git4notes/commit/7608ade443a3a8bd66a405628fd106ceb88fdb0b) was enough to address your (1) and (2) rules. I have tried this solution earlier but it did not work: probably something has been fixed since then, so updating state files for each processed branch works now.
Since each revision(which can live in different bzr branches) is imported only once now, we do not need your (3), because such situation is not possible. E.g.,  if two branches A and B have an identical revision R  with nicks nickA and nickB, and we import A branch first, B branch in git will reuse already imported A's revision with annotation "Bzr-Reference: nickA rR".
I performed tests, importing all official Squid branches (11) first  and all Measurement factory (29) branches then. After that, I ran our "compare_branches.sh" for MF branches.
BTW, during testing I found a problem that sometimes a bzr branch is not imported, because all its revision have been already imported from other branches. I added a [helper script](https://github.com/measurement-factory/bzr2git4notes/commit/ab38966d54c921a296919f17b169dfe8fd20562c) to analyse this situation. It is  separate issue, though.

**Comment by rousskov** [comment]
> if two branches A and B have an identical revision R with nicks nickA and nickB, and we import A branch first, B branch in git will reuse already imported A's revision with annotation "Bzr-Reference: nickA rR".

... which makes branch B comment references to that shared commit using to B-specific revision numbers unresolvable because the reader cannot know which git commit that "B rY" reference corresponds to. My (3) is needed to address that problem, but, for the reasons already discussed, I do not think we should address it.

**Comment by eduard-bagdasaryan** [comment]
> which makes branch B comment references to that shared commit using to B-specific revision numbers unresolvable because the reader cannot know which git commit that "B rY" reference corresponds to.

For 'plain' revisions I think that it is possible to identify a revision by its reference, because each git branch will get the same revision id numeration as its parent bzr branch had. Of course some of these revisions will differ with their branch nicks, but that is not so important, I think.

**Comment by rousskov** [comment]
Yes, the problem only exists for merged (a.k.a. dotted or not plain) branch commits.
